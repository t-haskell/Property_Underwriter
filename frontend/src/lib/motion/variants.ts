import type { Transition, Variants } from 'framer-motion';
import { distances, durations, easings } from './tokens';

const defaultTransition: Transition = {
  duration: durations.m,
  ease: easings.standard,
};

export const fade: Variants = {
  hidden: { opacity: 0 },
  show: { opacity: 1, transition: defaultTransition },
  exit: { opacity: 0, transition: { ...defaultTransition, duration: durations.s } },
};

export const rise: Variants = {
  hidden: { opacity: 0, y: distances.small },
  show: {
    opacity: 1,
    y: 0,
    transition: {
      duration: durations.m,
      ease: easings.entrance,
    },
  },
  exit: {
    opacity: 0,
    y: distances.tiny,
    transition: { duration: durations.s, ease: easings.exit },
  },
};

export const scale = (initialScale = 0.96): Variants => ({
  hidden: { opacity: 0, scale: initialScale },
  show: {
    opacity: 1,
    scale: 1,
    transition: {
      duration: durations.m,
      ease: easings.entrance,
    },
  },
  exit: {
    opacity: 0,
    scale: initialScale - 0.02,
    transition: { duration: durations.s, ease: easings.exit },
  },
});

type SlideDirection = 'up' | 'down' | 'left' | 'right';

export const slide = (
  direction: SlideDirection,
  distance: number = distances.small,
): Variants => {
  const axis = direction === 'left' || direction === 'right' ? 'x' : 'y';
  const multiplier = direction === 'left' || direction === 'up' ? -1 : 1;

  return {
    hidden: {
      opacity: 0,
      [axis]: multiplier * distance,
    },
    show: {
      opacity: 1,
      [axis]: 0,
      transition: {
        duration: durations.m,
        ease: easings.entrance,
      },
    },
    exit: {
      opacity: 0,
      [axis]: multiplier * (distance / 2),
      transition: { duration: durations.s, ease: easings.exit },
    },
  } as Variants;
};

export const staggerContainer = (
  stagger = 0.08,
  delayChildren = 0.04,
): Variants => ({
  hidden: {},
  show: {
    transition: {
      staggerChildren: stagger,
      delayChildren,
    },
  },
});

export const springyEnter: Variants = {
  hidden: { opacity: 0, scale: 0.92, y: distances.tiny },
  show: {
    opacity: 1,
    scale: 1,
    y: 0,
    transition: {
      type: 'spring',
      stiffness: easings.spring.stiffness,
      damping: easings.spring.damping,
      mass: easings.spring.mass,
    },
  },
  exit: {
    opacity: 0,
    scale: 0.96,
    y: distances.micro,
    transition: { duration: durations.s, ease: easings.exit },
  },
};
