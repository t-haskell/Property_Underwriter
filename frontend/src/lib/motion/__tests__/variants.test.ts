import { describe, expect, it } from 'vitest';
import { fade, rise, scale, slide, springyEnter, staggerContainer } from '../variants';

const snapshot = (value: unknown) => expect(value).toMatchSnapshot();

describe('motion variants', () => {
  it('matches expected fade blueprint', () => {
    snapshot(fade);
  });

  it('matches expected rise blueprint', () => {
    snapshot(rise);
  });

  it('provides slide variants per direction', () => {
    snapshot(slide('up', 16));
    snapshot(slide('right', 20));
  });

  it('creates springy entrance variant', () => {
    snapshot(springyEnter);
  });

  it('stagger container scales with delay', () => {
    snapshot(staggerContainer(0.08, 0.12));
  });

  it('scale variant respects provided factor', () => {
    snapshot(scale(0.9));
  });
});
