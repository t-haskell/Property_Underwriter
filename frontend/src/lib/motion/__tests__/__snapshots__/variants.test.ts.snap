// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`motion variants > creates springy entrance variant 1`] = `
{
  "exit": {
    "opacity": 0,
    "scale": 0.96,
    "transition": {
      "duration": 0.2,
      "ease": [
        0.4,
        0,
        1,
        1,
      ],
    },
    "y": 4,
  },
  "hidden": {
    "opacity": 0,
    "scale": 0.92,
    "y": 8,
  },
  "show": {
    "opacity": 1,
    "scale": 1,
    "transition": {
      "damping": 24,
      "mass": 1,
      "stiffness": 180,
      "type": "spring",
    },
    "y": 0,
  },
}
`;

exports[`motion variants > matches expected fade blueprint 1`] = `
{
  "exit": {
    "opacity": 0,
    "transition": {
      "duration": 0.2,
      "ease": [
        0.2,
        0,
        0,
        1,
      ],
    },
  },
  "hidden": {
    "opacity": 0,
  },
  "show": {
    "opacity": 1,
    "transition": {
      "duration": 0.36,
      "ease": [
        0.2,
        0,
        0,
        1,
      ],
    },
  },
}
`;

exports[`motion variants > matches expected rise blueprint 1`] = `
{
  "exit": {
    "opacity": 0,
    "transition": {
      "duration": 0.2,
      "ease": [
        0.4,
        0,
        1,
        1,
      ],
    },
    "y": 8,
  },
  "hidden": {
    "opacity": 0,
    "y": 16,
  },
  "show": {
    "opacity": 1,
    "transition": {
      "duration": 0.36,
      "ease": [
        0.16,
        1,
        0.3,
        1,
      ],
    },
    "y": 0,
  },
}
`;

exports[`motion variants > provides slide variants per direction 1`] = `
{
  "exit": {
    "opacity": 0,
    "transition": {
      "duration": 0.2,
      "ease": [
        0.4,
        0,
        1,
        1,
      ],
    },
    "y": -8,
  },
  "hidden": {
    "opacity": 0,
    "y": -16,
  },
  "show": {
    "opacity": 1,
    "transition": {
      "duration": 0.36,
      "ease": [
        0.16,
        1,
        0.3,
        1,
      ],
    },
    "y": 0,
  },
}
`;

exports[`motion variants > provides slide variants per direction 2`] = `
{
  "exit": {
    "opacity": 0,
    "transition": {
      "duration": 0.2,
      "ease": [
        0.4,
        0,
        1,
        1,
      ],
    },
    "x": 10,
  },
  "hidden": {
    "opacity": 0,
    "x": 20,
  },
  "show": {
    "opacity": 1,
    "transition": {
      "duration": 0.36,
      "ease": [
        0.16,
        1,
        0.3,
        1,
      ],
    },
    "x": 0,
  },
}
`;

exports[`motion variants > scale variant respects provided factor 1`] = `
{
  "exit": {
    "opacity": 0,
    "scale": 0.88,
    "transition": {
      "duration": 0.2,
      "ease": [
        0.4,
        0,
        1,
        1,
      ],
    },
  },
  "hidden": {
    "opacity": 0,
    "scale": 0.9,
  },
  "show": {
    "opacity": 1,
    "scale": 1,
    "transition": {
      "duration": 0.36,
      "ease": [
        0.16,
        1,
        0.3,
        1,
      ],
    },
  },
}
`;

exports[`motion variants > stagger container scales with delay 1`] = `
{
  "hidden": {},
  "show": {
    "transition": {
      "delayChildren": 0.12,
      "staggerChildren": 0.08,
    },
  },
}
`;
